### PR TITLE
Update Route API

### PR DESCRIPTION
**Pull request by marcogschmidt**
Update the Routing API to allow directing traffic directly to a Kubernetes service.

### CODE REVIEW

File: projects/gloo/api/v1/proxy.proto [hunk 1]
```diff
@@ -260,10 +260,25 @@ message RouteAction {
...
+// Destinations define routable destinations for proxied requests.
 message Destination {
-    // The upstream to route requests to
-    core.solo.io.ResourceRef upstream = 1 [(gogoproto.nullable) = false];
+
+    //  The type of the destination
+    oneof destination_type {
+
+        // Route requests to a Gloo upstream
+        UpstreamDestination upstream = 10;
```
Comment (yuval-k) [inline projects/gloo/api/v1/proxy.proto]: not sure we should break backwards compatibility here. can we leave this as core.solo.io.ResourceRef?

**Comment by marcogschmidt** [comment]
Just put this out here to get feedback. I feel this change cleaner, but would break the API. As an alternative, we could just change one field of the `Destination` message, like this:
```proto
// Destinations define routable destinations for proxied requests
message Destination {

    oneof destination_type {
        core.solo.io.ResourceRef upstream = 1;

        // Route requests to a kubernetes service
        ServiceDestination service = 11;
    }
    DestinationSpec destination_spec = 2;

    Subset subset = 3;
}
```

This way the change would be backwards compatible from the user's perspective, even though `DestinationSpec` and `Subset` currently are upstream-specific.

I lean towards the first option, but I don't want to discount disruptions to current users. Thoughts?  @ilackarms @yuval-k @rickducott

File: projects/gloo/api/v1/proxy.proto (hunk 1 above)
Comment (marcogschmidt) [reply projects/gloo/api/v1/proxy.proto]: You were fast in commenting! See my comment below. I am fine if you think breaking BC is an overkill here.

**Comment by rickducott** [comment]
I don't see any need to couple the break with the change -> `one_of` (I think you are advocating for it just because we're changing that API now and it is a better end state, but it would be roughly the same amount of work if we do the break now vs. later). 

When we do an API breaking release, we will want to think harder about how we're updating CRDs, supporting upgrade paths, and generally doing things that reduce the pain (communication, batching breaks, one or more major releases where the old API is functional but deprecated, etc).

**Comment by marcogschmidt** [comment]
@rickducott That makes a lot of sense. I was failing to see that the change and the API break could and should occur at different times. I'll go with the non-breaking change. Thanks for the feedback!

File: projects/gloo/api/v1/proxy.proto [hunk 2]
```diff
@@ -276,16 +285,25 @@ message Destination {
     Subset subset = 3;
 }
 
+// Identifies a port on a kubernetes service to route traffic to.
+message ServiceDestination {
+
+    // The target service
+    core.solo.io.ResourceRef metadata = 1 [(gogoproto.nullable) = false];
```
Comment (yuval-k) [inline projects/gloo/api/v1/proxy.proto]: metadata is a bit unclear to the average kube user. maybe `ref` or `service_ref`?

Comment (marcogschmidt) [reply projects/gloo/api/v1/proxy.proto]: I had it called `ref` in a previous commit, not sure why I changed it back. I'll change it back.

**Pull request merged**